src/Sparrow.Server/Allocators.Pool.FixedSize.cs

```diff
@@ -0,0 +1,270 @@
+            allocator.TotalAllocated += ptr.Size;
+
+            return new Pointer(ptr.Address, size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Release(ref FixedSizePoolAllocator<TOptions> allocator, ref Pointer ptr)
+        {
+            if (ptr.Size != allocator._options.MaxBlockSize)
+                ThrowMemoryDoesNotBelongToAllocator(ptr);
+
+            if (allocator.InUse > allocator._options.MaxPoolSizeInBytes || Allocator.LowMemoryFlag.IsRaised())
+                goto UnlikelyRelease;
+
+            int originalSize = ptr.Size;
+
+            var section = allocator._freed;
+            if (section.IsValid)
+            {
+                // Copy the section pointer that is already freed to the current memory. 
+                *(BlockPointer*)ptr.Address = section;
+            }
+            else
+            {
+                // Put a copy of the currently released memory block on the front. 
+                *(BlockPointer*)ptr.Address = new BlockPointer();
+            }
+
+            // Put a copy of the currently released memory block on the front. 
+            allocator._freed = ptr;
+            allocator.InUse -= originalSize;
+
+            ptr = new Pointer(); // Nullify the pointer
+            return;
+
+        UnlikelyRelease:
+            // This should be an unlikely case, when you are running out of memory or over allocated,
+            // all performance guarantees go down the drain. So we dont care if we hit expensive calls
+            // that allows us to release some steam at the cost of hitting cold code. 
+            // https://github.com/dotnet/coreclr/issues/6024
+
+            allocator.InUse -= ptr.Size;
+            allocator.Allocated -= ptr.Size;
+
+            Pointer nakedPtr = new Pointer(ptr.Address, ptr.Size);
+            allocator._internalAllocator.Release(ref nakedPtr);
+        }
+
+        private void ThrowMemoryDoesNotBelongToAllocator(in Pointer ptr)
+        {
+            throw new InvalidOperationException($"The memory pointer {ptr.Describe()} does not belong to this allocator.");
+        }
+
+        private void ThrowOnlyFixedSizeMemoryCanBeRequested()
+        {
+            throw new InvalidOperationException($"The memory size requested is not supported by this allocator. You should use {this._options.MaxBlockSize} instead.");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Renew(ref FixedSizePoolAllocator<TOptions> allocator)
+        {
+            if (Allocator.LowMemoryFlag.IsRaised())
+            {
+                ReleaseMemoryPool(ref allocator);
+                allocator.Allocated = 0;
+            }                
+
+            allocator._internalAllocator.Renew();
+            allocator.InUse = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Reset(ref FixedSizePoolAllocator<TOptions> allocator)
```

Comment:
I don't understand the usage of reset & renew in this PR

Reply:
I will add some information about it. But the concept is the following. 
Every allocator, when asked to reset, will get rid of as much memory as it possibly can. Allocators when asked to `Reset` wont try to speculatively hoard on memory, they will release as much as they can to the parent allocator. On the other hand, calling Renew may be a 'no-op' in the sense that the allocator is free to speculatively hoard the memory for its own use. For example in the case of the `PoolAllocator` when asked to Reset it will `clean-up` its internal storage and release the pools to the parent allocator or release the memory if it has ownership. While when asked to `Renew` it will only release the internal store if there is a Low Memory Flag activated and keep the memory in hot standby to use on the next cycle. 

That is the reason why we track not only the memory that the allocator has given to the callers (which could hint on internal fragmentation) but also the total allocated memory on stores. If those numbers are big, it would make sense to issue a `Reset`, or for example after an indexing call, etc... While for most cases you will end up only calling `Renew`.

Reply:
In that case, let's give these method different names, because they are conflicting with the existing terminology in the codebase.

Another issue is who call them, and when? 
When is Reset and when is Renew called?

Reply:
Makes a lot of sense to change the same to avoid mistakes. Any suggestion there?
Those are called by the owner, which in our case and on our current design is `JsonOperationContext` and the indexing threads... I am not entirely sure we should keep that that way, even though it makes sense to have a single `OperationContext` manager which will decide based on a sensible heuristic if to call one or the other.

Reply:
`Renew` -> `ResetState` and `Reset` -> `ReleaseResources` ?

And who calls what is critical to the design.
We currently checkout a context, use it and return it.
Who does what in this case?

Reply:
Not fond of those names because `Reset` and `ResetState` can become false friends, but not as bad as `Reset` and `Renew` semantically speaking.

Reply:
In the process of returning it, the general manager should decide before checkout of the context what to do with the allocator... a lease on the allocators themselves is not a bad idea either.

Reply:
That is too generic. Consider the scenarios we have.
For indexing, we know if we'll have a new indexing round or if we are going to rest, so that make sense.
But for requests, we have no such information, how do we deal with it then?

Reply:
For that how old or how much memory it has allocated makes sense as a possible heuristic.